### PR TITLE
LibWeb: Implement the start of the Navigation API

### PR DESCRIPTION
**Pull request by ADKaster**
This API is how JavaScript can manipulate the new Navigable concepts
directly. It includes an interface to SessionHistoryEntries, called
NaviagtionHistoryEntry, A whole host of new navigation Events, and a new
Window property named ``navigation``.

This patch includes a full implementation of NavigationHistoryEntry and
the NavigationCurrentEntryChangeEvent which is fired when updating the
current history entry. It is missing most of the interesting algoritms
on Navigation that do the actual navigation steps, and call into the
currently WIP navigable AOs.

Some cleanup of SessionHistoryEntry is also possible, as we never
actually implemented any functionality for the serialized_data field.

---

Also Implement more of the shared history push/replace state steps

Add the seralization and URL validation steps, but skip the actual
navigation for now. This might cause more pages to throw exceptions
when trying to push state that contains objects that we don't know how
to serialize.

### CODE REVIEW

**Review by shannonbooth — APPROVED** [review]
very cool!

File: Userland/Libraries/LibWeb/HTML/NavigationType.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    Replace,
+    Reload,
+    Traverse,
+};
+
+inline String idl_enum_to_string(NavigationType value)
+{
+    switch (value) {
+    case NavigationType::Push:
+        return MUST(String::from_utf8("Push"sv));
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/NavigationType.h:24]: Oops, my comment didn't drop, I meant to say here - maybe this can just return `"Push"_string`

Comment (ADKaster) [reply Userland/Libraries/LibWeb/HTML/NavigationType.h:24]: Ah, good catch! I'll fix that one up in the morning

**Comment by ADKaster** [comment]
Splitting this into more commits

File: Userland/Libraries/LibWeb/HTML/Navigation.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    // Each Navigation has an associated entry list, a list of NavigationHistoryEntry objects, initially empty.
+    Vector<JS::NonnullGCPtr<NavigationHistoryEntry>> m_entry_list;
+
+    // https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-current-entry-index
+    // Each Navigation has an associated current entry index, an integer, initially −1.
+    i64 m_current_entry_index { -1 };
+
+    // https://html.spec.whatwg.org/multipage/nav-history-apis.html#concept-navigation-transition
+    // Each Navigation has a transition, which is a NavigationTransition or null, initially null.
+    JS::GCPtr<NavigationTransition> m_transition { nullptr };
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/HTML/Navigation.h:94]: should we visit `m_transition` in `Navigation::visit_edges()`?

Comment (ADKaster) [reply Userland/Libraries/LibWeb/HTML/Navigation.h:94]: Ah! nice catch. Fixed that one up.

**Pull request merged**